### PR TITLE
GROOVY-6992: Added support for collecting over sets

### PR DESCRIPTION
**Pull request by mperry**
Pull request for https://jira.codehaus.org/browse/GROOVY-6992.

Collecting over sets without passing in a collection to add to returns a list.  We really want it to return a set.  

I added support for this to DefaultGroovyMethods for a set using the default set of LinkedHashSet.  I noticed this was used as the default.  It worked for HashSet too, but I have kept the current default.  I initially tried TreeSet, but this class is broken in Java and should always require an ordering operation for elements.

A warning, this may break existing code that relies on collecting over a set returning a list.  For this behaviour, toList() can be called either before or after the collect method.

### CODE REVIEW

**Comment by eric-milles** [comment]
As of Groovy 4, the `collect` variants that accept a collector propagate the collector's type.  For example:
```groovy
@groovy.transform.TypeChecked
void test() {
  def one_two_three = [1,2,3]
  @groovy.transform.ASTTest({
    println node.getNodeMetaData(org.codehaus.groovy.transform.stc.StaticTypesMarker.INFERRED_TYPE)
  })
  def two_times = one_two_three.collect(new HashSet<Number>()) { it * 2 }
}

test()
```

This script prints "java.util.HashSet<java.lang.Number>" for the inferred type of `two_times`.  This lets you control the type and carry on with that type instead of having to cast from `Collection` as before.

**Comment by mperry** [comment]
I found your example hard to follow @eric-milles, but it is promising if collect preserves the type.
